course02/homework02/danielellis/Buffer.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <deque>
+#include <chrono>
+#include <vector>
+#include <stdexcept>
+
+struct WindowConfig
+{
+	std::chrono::seconds mDuration = std::chrono::seconds(10);
+	int mMaxNumMessages = 5;
+};
+
+template<class Message>
+struct TimestampedMessage
+{
+	std::chrono::time_point<std::chrono::steady_clock> mTimestamp;
+	Message mMessage;
+};
+
+template<class Message>
+class Buffer
+{
+public:
+	using value_type = Message;
+
+	Buffer(const WindowConfig& conf);
+	void TryToAddMessage(const Message& message);
+	std::vector<Message> DumpMessages() const;
+
+private:
+	bool CanAddMessage(std::chrono::time_point<std::chrono::steady_clock> currentTime) const;
+
+	WindowConfig mWindowConfig;
+	std::deque<TimestampedMessage<Message> > mTimestampedMessages;
```

Comment:
Why not, although it is not mandatory to copy the messages themselves if you only want to throttle. You only need the timestamps. Depending on the type, the message can be quite big and expensive to copy.